database/raft/raft.go

```diff
@@ -0,0 +1,935 @@
+}
+
+// ServeHTTP responds to raft consensus messages at /raft/x,
+// where x is any particular raft internode RPC.
+// When sv sends outgoing messages, it acts as an HTTP client
+// and sends requests to its peers at /raft/x.
+func (sv *Service) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	sv.mux.ServeHTTP(w, req)
+}
+
+func (sv *Service) serveMsg(w http.ResponseWriter, req *http.Request) {
+	b, err := ioutil.ReadAll(http.MaxBytesReader(w, req.Body, maxRaftReqSize))
+	if err != nil {
+		http.Error(w, "cannot read req: "+err.Error(), 400)
+		return
+	}
+	var m raftpb.Message
+	err = m.Unmarshal(b)
+	if err != nil {
+		http.Error(w, "cannot unmarshal: "+err.Error(), 400)
+		return
+	}
+	sv.raftNode.Step(req.Context(), m)
+}
+
+func (sv *Service) serveJoin(w http.ResponseWriter, req *http.Request) {
+	var x struct{ Addr string }
+	err := json.NewDecoder(req.Body).Decode(&x)
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	log.Printkv(req.Context(), "at", "join", "addr", x.Addr)
+
+	newID, err := sv.allocNodeID(req.Context())
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	log.Printkv(req.Context(), "at", "join-id", "addr", x.Addr, "id", newID)
+
+	err = sv.raftNode.ProposeConfChange(req.Context(), raftpb.ConfChange{
+		ID:      atomic.AddUint64(&sv.confChangeID, 1),
+		Type:    raftpb.ConfChangeAddNode,
+		NodeID:  newID,
+		Context: []byte(x.Addr),
+	})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	log.Printkv(req.Context(), "at", "join-done", "addr", x.Addr, "id", newID)
+
+	snap := sv.getSnapshot()
+	snapData, err := encodeSnapshot(snap)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	json.NewEncoder(w).Encode(nodeJoin{newID, snapData})
+}
+
+// join attempts to join the cluster.
+// It requests an existing member to propose a configuration change
+// adding the local process as a new member, then retrieves its new ID
+// and a snapshot of the cluster state and applies it to sv.
+// It also sets sv.id.
+func (sv *Service) join(addr, baseURL string) error {
+	reqURL := strings.TrimRight(baseURL, "/") + "/raft/join"
+	b, _ := json.Marshal(struct{ Addr string }{addr})
+	resp, err := http.Post(reqURL, contentType, bytes.NewReader(b))
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	var x nodeJoin
+	err = json.NewDecoder(resp.Body).Decode(&x)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	sv.id = x.ID
+	var raftSnap raftpb.Snapshot
+	err = decodeSnapshot(x.Snap, &raftSnap)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	ctx := context.Background()
+	err = sv.raftStorage.ApplySnapshot(raftSnap)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	if raft.IsEmptySnap(raftSnap) {
+		log.Printkv(ctx, "at", "joined", "appliedindex", 0, "msg", "(empty snap)")
+	} else {
+		err = sv.state.RestoreSnapshot(raftSnap.Data, raftSnap.Metadata.Index)
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		sv.stateMu.Lock()
```

Comment:
Acquiring here isn't necessary because the http listener hasn't started yet, so this "happens before" any of the handlers can start. This is also why we don't hold the lock when calling RestoreSnapshot just above. Basically, we have a "setup" phase and then an "online" phase, and we don't take the lock during the setup phase because there's no concurrency in play.

I am ok with taking the lock on this line, but if we do it, then we should also do it consistently throughout the setup phase whenever we're modifying state fields.